mock-builder/src/location.rs

```diff
@@ -32,42 +50,104 @@ impl FunctionLocation {
-		let path = path
-			.strip_prefix('<')
-			.map(|trait_path| trait_path.split_once(" as").expect("always ' as'").0)
-			.unwrap_or(path);
+		let (path, name) = self.location.rsplit_once("::").expect("always '::'");
+		let (path, trait_info) = match path.strip_prefix('<') {
+			Some(struct_as_trait_path) => {
+				let struct_path = struct_as_trait_path
+					.split_once(" as")
+					.expect("always ' as'")
+					.0;
+
+				let trait_name = struct_as_trait_path
+					.rsplit_once("::")
+					.expect("always '::'")
+					.1
+					.strip_suffix('>')
+					.unwrap();
 
-		Self(format!("{path}::{name}"))
+				(struct_path, Some(trait_name.to_owned()))
+			}
+			None => (path, None),
+		};
+
+		Self {
+			location: format!("{path}::{name}"),
+			trait_info,
+		}
 	}
 
 	/// Remove the prefix from the function name.
 	pub fn strip_name_prefix(self, prefix: &str) -> Self {
-		let (path, name) = self.0.rsplit_once("::").expect("always ::");
+		let (path, name) = self.location.rsplit_once("::").expect("always ::");
 		let name = name.strip_prefix(prefix).unwrap_or_else(|| {
 			panic!(
 				"Function '{name}' should have a '{prefix}' prefix. Location: {}",
-				self.0
+				self.location
 			)
 		});
 
-		Self(format!("{path}::{name}"))
+		Self {
+			location: format!("{path}::{name}"),
+			trait_info: self.trait_info,
+		}
+	}
+
+	/// Remove the trait name from the function name and add such information to
+	/// the location. The location is expected to have the following structure:
+	/// `<path>::<TraitInfo>_<name>`
+	pub fn assimilate_trait_prefix(self) -> Self {
+		let (path, name) = self.location.rsplit_once("::").expect("always ::");
+		let (trait_info, name) = match name.chars().next().unwrap().is_uppercase() {
+			true => {
+				let (trait_info, name) = name.split_once('_').expect("always '_' after trait name");
+				(Some(trait_info.to_owned()), name)
+			}
+			false => (None, name),
+		};
+
+		Self {
+			location: format!("{path}::{name}"),
+			trait_info,
+		}
 	}
 
 	/// Add a representation of the function input and output types
 	pub fn append_type_signature<I, O>(self) -> Self {
-		Self(format!("{}:{}", self.0, TypeSignature::new::<I, O>()))
+		Self {
+			location: format!("{}:{}", self.location, TypeSignature::new::<I, O>()),
+			trait_info: self.trait_info,
+		}
 	}
 
 	/// Generate a hash of the location
-	pub fn hash<Hasher: StorageHasher>(&self) -> Hasher::Output {
-		Hasher::hash(self.0.as_bytes())
+	pub fn hash<Hasher: StorageHasher>(&self, trait_info: TraitInfo) -> Hasher::Output {
+		let string = match trait_info {
+			TraitInfo::Yes => {
+				let trait_info = self
+					.trait_info
+					.as_ref()
+					.expect("Location must have trait info");
+				format!("{}{}", self.location, trait_info)
+			}
+			TraitInfo::No => self.location.clone(),
+			TraitInfo::Whatever => {
+				let trait_info = self.trait_info.clone().unwrap_or_default();
+				format!("{}{}", self.location, trait_info)
+			}
+		};
+
+		dbg!(trait_info, &string);
```

Comment:
I suppose this is a remnant from debugging which we don't want to merge?

Reply:
Thanks for your eyes!